sycl/source/detail/device_impl.cpp

```diff
@@ -342,6 +342,17 @@ bool device_impl::isAssertFailSupported() const {
   return MIsAssertFailSupported;
 }
 
+std::string device_impl::getDeviceName() const {
+  if (MDeviceName.empty()) {
```

Comment:
Race condition will occur here, and that's why I double check it after mutex is acquired. As for the constructors, this might have been a good idea, but we have plenty of those, and I'm worried eventually we'll screw things up and report incorrect values.

Reply:
On the second thought, what I want to do here is [`std::call_once`](https://en.cppreference.com/w/cpp/thread/call_once)